Add human notes to a referral

patinipat opened the pull request:
# Context

<!-- Is there a Trello ticket you can link to? -->
<!-- Do you need to add any environment variables? -->
<!-- Is an ADR required? An ADR should be added if this PR introduces a change to the architecture. -->

# Changes in this PR

Add a create assessment note form on the review and assess show page
Notes are displayed using the moj Timeline component


## Screenshots of UI changes

### Before
![localhost_3000_review-and-assess_df0e4635-b71d-4ad2-8191-d74b1bcb27ad_df0e4635-b71d-4ad2-8191-d74b1bcb27ad_summary_success=false](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/65b4c491-0149-4a03-8790-5f6a74b4be88)

### After
![localhost_3000_review-and-assess_df0e4635-b71d-4ad2-8191-d74b1bcb27ad_df0e4635-b71d-4ad2-8191-d74b1bcb27ad_summary_success=true](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/94137563/c7b41934-69d1-41ff-8a6f-818e37cb7a49)

# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

bryangaledxw commented:
Giving this a preliminary pass, all looks good. Happy with the commit structure, building up the layers. I think maybe it'd be worth putting the new util method (and its inclusion in nunjucks setup) in it's own commit before the controller changes, but that's a minor thing. Some other things I spotted:

* Should `text: note.message` in assessment utils be `html: formatLines(note.message)`? Just want to make sure notes with line breaks are handled properly. Suggest it's worth using `faker.lorem.lines()` in the factory to exercise this
* Do you need any error handling in the new controller method. Does/should the API allow empty notes? You'd need to do the same style of error handling as in `PremisesController.create()` and add a new message to `errors.json`

libuk reviewed (commented):
Looks good. I agree with @bryangaledxw that the utils change could be pulled out into a separate commit, but shouldn't be a blocker. I'll defer to him for approval as there may be some context I'm missing.

patinipat committed:
Add helper function timelineItems()

This helper function transforms assessment notes to a format accepted by the MoJ Timeline component, sorted for presentation in reverse-chronological order.

As appears to be standard MoJ practice for user names[1], we use convertToTitleCase() to format the name for display.

Though the API currently only supports user notes, we expect soon it will return non-user notes, so we ensure we do not depend on the presence of createdByUserName on our note object.

[1] https://github.com/ministryofjustice/hmpps-template-typescript/blob/fc402c224eee3eeec0f1f56e1b92e452648be1f7/server/services/userService.ts#L14

patinipat commented:
Extract the new helper function in assessmentUtils in its own commit

bryangaledxw commented on server/views/temporary-accommodation/assessments/show.njk:
```diff
@@ -66,6 +69,38 @@
+            classes: "govuk-label--m referral-note"
+          }
+        })
+        }}
```
Suggest use taTextarea here instead, it'll save some repetition. And use the same naming as the field in the API. If this is "message" rather than "notes", everything will link up by magic (ie. `catchValidationErrorOrPropogate`) if there's an error from the API - this'll be be highlighted in red, and the errors at the top of the page will link to here

patinipat replied:
:recycle:

bryangaledxw commented on server/utils/nunjucksSetup.ts:
```diff
@@ -92,6 +92,7 @@ export default function nunjucksSetup(app: express.Express, path: pathModule.Pla
   njkEnv.addGlobal('dateFieldValues', dateFieldValues)
   njkEnv.addGlobal('formatDate', DateFormats.isoDateToUIDate)
   njkEnv.addGlobal('formatDateTime', DateFormats.isoDateTimeToUIDateTime)
+  njkEnv.addFilter('mojDate', DateFormats.isoDateTimeToUIDateTime)
```
So is this used internally by the timeline component?

patinipat replied:
yes https://github.com/ministryofjustice/moj-frontend/blob/main/src/moj/components/timeline/template.njk#L18

bryangaledxw commented on cypress_shared/pages/assess/show.ts:
```diff
@@ -52,4 +67,36 @@ export default class AssessmentShowPage extends Page {
+      .parent()
+      .parent()
+      .within(() => {
+        cy.get('.govuk-textarea').type(message)
```
Think this could be a call down to `completeTextInputByLabel` in the base class - `completeTextInputByLabel('Add a note', message)`. And getting more into personal preference, consider just doing something like `cy.get('button').contains('Save note').click()`

patinipat replied:
:recycle:

bryangaledxw reviewed (commented):
Gone through a bit more closely. I'll hold off regarding as done until you've had a chat with Nick about showing the name of the person who created the note, but this doesn't look far off, good stuff

patinipat commented:
Added support to display the name of the staff member who created the note. This relies on the `User` type being updated to provide the `name` property. This commit https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/pull/559/commits/678d2299e68e504bd32da39f36ebd13ac3bec44f is there in case we need to format the user.name has Im not sure what will format will be provided.

bryangaledxw committed:
Add MoJ filters to Nunjucks

We will later be using the MoJ timeline component, which has a dependence on a set of standard MoJ Nunjucks filters. We expose these filters to our templates, overriding the mojDate filter so that dates are formatted consistent with elsewhere in our UI. Unfortunately, these filters are not exposed in a way we can `import`, so we have to use `require` here

bryangaledxw committed:
Add createNote() to AssessmentService

We update the service to expose the AssessmentClient.createNote() method from our previous commit

bryangaledxw committed:
Add createNote() to AssessmentClient

We add a createNote() method to the AssessmentClient along with the new API path associated with it. We also stub this new endpoint for local development, and add Cypress tasks we will use later to mock this endpoint in integration tests

bryangaledxw committed:
Sort imports

bryangaledxw committed:
Update assessment test naming

Before expanding these tests, we update the current naming to follow project conventions

bryangaledxw committed:
Add notes to assessment summary page

We now use our timelineItems() helper function to display user notes on the assessment summary page

bryangaledxw committed:
Allow users to create a new note

We now add a note creation form the the assessment summary page, implemented using a new createNote() route on our AssessmentController. We also update our summary() route to forward any errors arising from note creation onto our template

creation

bryangaledxw committed:
Audit assessment note creation

Typically we use the redirect destination of a route handler to determine whether to audit a POST as having succeeded or failed. Since in both cases our AssessmentController.createNote() handler redirects to the same page, we need to add an otherwise unused query parameter to distinguish the two outcomes for our audit middleware

bryangaledxw committed:
Update assessment and related factories

We update our assessment factory to generate an assessment with user notes. To enable this we add a user note factory, along with a generic note factory for completeness

bryangaledxw committed:
Update assessment page constructors

As AssessmentSummaryPage will need to be aware of notes, we update the constructor to take the entire assessment, rather than isolated fields. We similarly update AssessmentFullPage, and add static visit() methods as per convention

bryangaledxw commented:
@libuk Ready for another look

libuk commented on wiremock/assessmentStubs.ts:
```diff
@@ -86,6 +87,19 @@ const assessmentStubs = [
           },
         },
+        {
+          priority: 99,
```
What does this priority key do out of interest?

bryangaledxw replied:
Generally when we're stubbing a POST endpoint, we'll have the success response that matches against a path, and some error responses that match against the path plus specific fields being missing. We want to make sure that the success response takes lower priority, otherwise that'd match first and we'd never see the error response in our wiremock API

Worth having a look at `errorStub()` and `getCombinations()`. We don't use the latter here as there's only one field (`message`), but generally we use the two together to generate error stubs for any combination of missing fields

libuk reviewed (approved):
Nice work 👍🏾
